contrib/endpoints/src/api_manager/firebase_rules/firebase_request.cc

```diff
@@ -0,0 +1,402 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+#include "contrib/endpoints/src/api_manager/firebase_rules/firebase_request.h"
+#include "contrib/endpoints/src/api_manager/utils/marshalling.h"
+#include "contrib/endpoints/src/api_manager/utils/url_util.h"
+#include "google/protobuf/util/message_differencer.h"
+
+#include <algorithm>
+#include <sstream>
+using ::google::api_manager::utils::Status;
+using ::google::api_manager::proto::TestRulesetResponse;
+using ::google::protobuf::util::MessageDifferencer;
+using ::google::protobuf::Map;
+
+namespace google {
+namespace api_manager {
+namespace firebase_rules {
+
+namespace {
+
+const char kToken[] = "token";
+const char kAuth[] = "auth";
+const char kPath[] = "path";
+const char kMethod[] = "method";
+const char kHttpGetMethod[] = "GET";
+const char kHttpPostMethod[] = "POST";
+const char kHttpHeadMethod[] = "HEAD";
+const char kHttpOptionsMethod[] = "OPTIONS";
+const char kHttpDeleteMethod[] = "DELETE";
+const char kFirebaseCreateMethod[] = "create";
+const char kFirebaseGetMethod[] = "get";
+const char kFirebaseDeleteMethod[] = "delete";
+const char kFirebaseUpdateMethod[] = "update";
+const char kV1[] = "/v1";
+const char kTestQuery[] = ":test?alt=json";
+
+void SetProtoValue(const std::string &key,
+                   const ::google::protobuf::Value &value,
+                   ::google::protobuf::Value *head) {
+  ::google::protobuf::Struct *s = head->mutable_struct_value();
+  Map<std::string, google::protobuf::Value> *fields = s->mutable_fields();
+  (*fields)[key] = value;
+}
+
+// Convert HTTP method to Firebase specific method.
+std::string GetOperation(const std::string &httpMethod) {
```

Comment:
Here is Google's c++ style guide's decision on global or static std::string objects. To my understanding, this seems conflicting advice. Let me know if we have decided not to follow this guideline and I am also curious as to why we have decided this is not the best thing to do ...

https://engdoc.corp.google.com/eng/doc/devguide/cpp/styleguide.shtml?cl=head#Static_and_Global_Variables